buildtest/utils/command.py

```diff
@@ -54,27 +116,26 @@ def execute(self):
         # Use updated command with executable and remainder (list)
         cmd = [executable] + args
 
-        # open the process for writing
-        process = subprocess.Popen(cmd, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
-        returncode = process.poll()
-
-        # Iterate through the output
-        while returncode is None:
-
-            out = self.decode(process.stdout.readline())
-            err = self.decode(process.stderr.readline())
-
-            # Append output and error
-            if out:
-                self.out.append(out)
-            if err:
-                self.err.append(out)
+        # Capturing provides temporary output and error files
+        with Capturing() as capture:
+            process = subprocess.Popen(
+                cmd,
+                stdout=capture.stdout,
+                stderr=capture.stderr,
+                universal_newlines=True,
+            )
             returncode = process.poll()
 
+            # Iterate through the output
+            while returncode is None:
+                returncode = process.poll()
+
         # Get the remainder of lines, add return code
-        out, err = process.communicate()
-        self.out += ["%s\n" % x for x in self.decode(out).split("\n") if x]
-        self.err += ["%s\n" % x for x in self.decode(err).split("\n") if x]
+        self.out += ["%s\n" % x for x in self.decode(capture.out).split("\n") if x]
```

Comment:
The output from ``capture.out`` is in string, I am wondering if it makes sense to return output as a ``str`` as pose to ``list`` when we return from method. 

I see the ``self.decode`` is going through each line and decode based on encoding, i am fine with this but maybe once ``self.out`` is a list to convert it back to a string. I can see when invoking ``BuildTestCommand`` to get the output and error and having to convert list to string everytime would be a bit annoying.

Reply:
It's no longer reading from stdout / stderr, but just reading the written file, so probably the decode isn't even necessary. This is also why the custom parsing (e.g., splitting by newline) is handled by the calling class - someone else using the Capturing() class might just want the entire blob of text read it, sans any additional processing. This is why I made this explicit choice.

Reply:
Yea i can myself preferring to read output as a text blob as pose to converting the list to string. Maybe just a thought if we can have a method in ``buildtest/utils/file.py`` let's say **list_to_string** that automatically converts a list to string. It can be as simple as 

```
def list_to_string(content):
  return ' '.join(content)
```

If we add this method i'd be fine with having BuildTestCommand return output as a list so i can convert it to a string.

Reason is i can see BuildTestCommand used in many places where we need output as a string. One example is in https://github.com/buildtesters/buildtest/blob/devel/buildtest/docgen/main.py  which uses BuildTestCommand to run examples and write output to a txt file. This helps create the output of the test automatically for documentation as pose to running each command manually. I found that as buildtest changes its behavior i'd like a script to rebuild all the doc examples.

Reply:
You could, but since this function is just one line, and abstracts away a relatively simple operation `content.split("\n")` I don't think this is a useful feature to add. the "split" function provided by the string class already handles this, and in one simple, clear command. The same is the case for join. :-1: from me.